operator/README.md

```diff
@@ -0,0 +1,173 @@
+
+#### DeploymentSpec
+
+|Field | type | Description | Default Value |
+|---|---|---|---|
+| driverName | string | Unique CSI driver name to use | `pmem-csi.intel.com` |
+| image | string | PMEM-CSI docker image name used for the deployment | `intel/pmem-csi-driver:canary` |
+| provisionerImage | string | [CSI provisioner](https://github.com/kubernetes-csi/external-provisioner) docker image name | `quay.io/k8scsi/csi-provisioner:v1.2.1` |
+| registrarImage | string | [CSI node driver registrar](https://github.com/kubernetes-csi/node-driver-registrar) docker image name | `quay.io/k8scsi/csi-node-driver-registrar:v1.1.0` |
+| pullPolicy | string | Docker image pull policy. either one of `Always`, `Never`, `IfNotPresent` | `IfNotPresent` |
+| logLevel | integer | PMEM-CSI driver logging level | 3 |
+| deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
+| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod |
+| nodeResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for the pods running on node(s) |
+| nodeSelector | string map | [Labels to use for selecting Nodes](../README.md#run-pmem-csi-on-kubernetes) on which PMEM-CSI driver should run. | `{ "storage": "pmem" }`
+
+#### DeploymentStatus
+
+A PMEM-CSI Deployment's `status` field is a `DeploymentStatus` object, which has
+a `phase` field. The phase of a Deployment is high-level summary of where the
+Deployment is in it's lifecycle.
+
+The possible `phase` values and their meaning are as below:
+
+| value | meaning |
+|---|---|---|
+| Pending | The `Deployment` has been accepted by the system, but waiting for one or more of it's pre-requisites to be completed. Such as waiting for required [TLS certificates](../README.md#security) to be generated |
+| Initializing | The required sub-resources of the `Deployment` are being initialized |
+| Running | All the required sub-resources of the `Deployment` are created |
+| Failed | For some reason the state of the `Deployment` failed and cannot be progressed |
+
+### TLS Certificate approval
+
+All PMEM-CSI driver communication is protected by mutual TLS. So the driver
+deployment needs to generate those certificates. The operator uses Certificate
+Authority(CA) provided by the Kubernetes cluster to get those certificates. On a
+new deployment, the operator creates Kubernetes `CertificateSigningRequest`
+objects and keeps the deployment in `pending` state till those CSRs get approved
+by an authorized user so that Kubernetes CA could generate the certificates.
+Once the CSRs are approved and certificates are ready, then the deployment
+process gets progressed.
+
+If the operator is configured with appropriate RBAC rules with enough
```

Comment:
Isn't that always going to be the case when deploying the driver as intended?

As it stands now, this sentence begs the question how an admin can configure this, which isn't explained.

Reply:
Yes, a valid question.

My idea was that we should make this certificate approval configurable so that admins can choose their own approval mechanisms. But how to is a bit unclear myself. I couldn't decide between 

    a) providing a command-line argument 
    b) missing RBAC(CSR and CSR Approval) rules. 

In the first case, we could provide an option something like `--run-csr-controller` where the operator manifest _must_ have needed RBAC rules otherwise we fail to start the operator. 

In the second case, the operator tries to start csr controller, and if it fails with lack of permissions then just ignore the error continue with no csr controller. In this case, the admin could drop CSR related rules from the provided default operator manifest.

So I just assumed the second option, though not implemented :)

Reply:
> In the second case, the operator tries to start csr controller, and if it fails with lack of permissions then just ignore the error continue with no csr controller.

Ignore silently or log as an error? If it's flagged as an error, then there handling the same in both cases.

If it's ignored silently, then accidentally missing RBAC rules would be hard to debug. Better always log this as an error.

A `--run-csr-controller` option might make sense if we can figure out a use cases for it. I wouldn't bother initially and always run it.